staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,290 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package events
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/clock"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	restclient "k8s.io/client-go/rest"
+
+	"os"
+
+	"k8s.io/api/events/v1beta1"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record/util"
+	"k8s.io/klog"
+)
+
+const maxTriesPerEvent = 12
+
+const maxQueuedEvents = 1000
+
+var defaultSleepDuration = 10 * time.Second
+
+type eventKey struct {
+	action              string
+	reason              string
+	reportingController string
+	reportingInstance   string
+	regarding           corev1.ObjectReference
+	related             *corev1.ObjectReference
+}
+
+type eventBroadcasterImpl struct {
+	*watch.Broadcaster
+	eventCache    map[eventKey]*v1beta1.Event
+	mu            sync.RWMutex
+	sleepDuration time.Duration
+	sink          EventSink
+}
+
+// NewBroadcaster Creates a new event broadcaster.
+func NewBroadcaster(sink EventSink) EventBroadcaster {
+	return &eventBroadcasterImpl{
+		Broadcaster:   watch.NewBroadcaster(maxQueuedEvents, watch.DropIfChannelFull),
+		eventCache:    map[eventKey]*v1beta1.Event{},
+		sleepDuration: defaultSleepDuration,
+		sink:          sink,
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) eventThreshold() {
+	eventBroadcaster.mu.RLock()
+	defer eventBroadcaster.mu.RUnlock()
+	for _, event := range eventBroadcaster.eventCache {
+
+		if event.Series != nil {
+			recordEvent(eventBroadcaster.sink, event, true)
+		}
+
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) checkEndingSeries() {
+	eventBroadcaster.mu.Lock()
+	defer eventBroadcaster.mu.Unlock()
+	for isomorphicKey, event := range eventBroadcaster.eventCache {
+		eventSerie := event.Series
+		if eventSerie != nil {
+
+			if eventSerie.LastObservedTime.Time.Add(6 * time.Minute).Before(time.Now()) {
+
+				if recordEvent(eventBroadcaster.sink, event, true) {
+					delete(eventBroadcaster.eventCache, isomorphicKey)
+				}
+			}
+		}
+
+	}
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (eventBroadcaster *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	return &recorderImpl{scheme, reportingController, reportingInstance, eventBroadcaster.Broadcaster, clock.RealClock{}}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) recordToSink(event *v1beta1.Event, randGen *rand.Rand, clock clock.Clock) {
```

Comment:
randGen and clock seem like things that should be fields in eventBroadcaster, not passed to every call of this.